File: Python/Medium/3546 Equal Sum Grid Partition I.py
class Solution:
    def canPartitionGrid(self, grid: List[List[int]]) -> bool:
        m, n = len(grid), len(grid[0])
        total_sum = sum(sum(row) for row in grid)
        
        if total_sum % 2 != 0:
            return False
        
        target = total_sum // 2
        
        # Check Horizontal Cuts
        current_row_sum = 0
        for i in range(m - 1):
            current_row_sum += sum(grid[i])
            if current_row_sum == target:
                return True
                
        # Check Vertical Cuts
        current_col_sum = 0
        for j in range(n - 1):
            for i in range(m):
                current_col_sum += grid[i][j]
            if current_col_sum == target:
                return True
                
        return False

File: C/Medium/3546 Equal Sum Grid Partition I.c
#include <stdbool.h>

bool canPartitionGrid(int** grid, int gridSize, int* gridColSize) {
    int m = gridSize;
    int n = gridColSize[0];
    long long totalSum = 0;

    for (int i = 0; i < m; i++) {
        for (int j = 0; j < n; j++) {
            totalSum += grid[i][j];
        }
    }

    if (totalSum % 2 != 0) return false;
    long long target = totalSum / 2;

    // Check Horizontal Cuts
    long long currentHeightSum = 0;
    for (int i = 0; i < m - 1; i++) {
        for (int j = 0; j < n; j++) {
            currentHeightSum += grid[i][j];
        }
        if (currentHeightSum == target) return true;
    }

    // Check Vertical Cuts
    long long currentWidthSum = 0;
    for (int j = 0; j < n - 1; j++) {
        for (int i = 0; i < m; i++) {
            currentWidthSum += grid[i][j];
        }
        if (currentWidthSum == target) return true;
    }

    return false;
}
